### DISCUSSION TITLE
학습 플랜 디비  및 Entity 설계

### DISCUSSION THREAD

**Discussion by InHyeok-J** [Jabiseo/Jabiseo-Backend#64] [backend-tech]
### 초기 기획

기능적 요구사항

- **학습 플랜**은 사용자가 학습중인 자격증 마다 하나만 진행할 수 있다.
- **학습 플랜 목록**은 미리 정해진 값이며, 일간 및 주간 단위로 설정할 수 있다.
- 학습 플랜의 목표는 **학습 시간**이나 **문제 풀이 수**로 정의된다.
- 학습 플랜을 조회할 때, 달성률이 달력의 월 단위로 체크된다.
- 플랜 수정이 가능하다.
- **플랜 수정과 이력관리 기능 추가**
    - 플랜 아이템의 수정은 **목표 값을 바꾸거나**, 기존 ENUM 값 중 **새로운 플랜 아이템을 추가**하거나 **기존 플랜 아이템을 삭제**할 수 있어야 한다.
    - 플랜 아이템들이 수정되는 경우 과거의 플랜 달성률을 유지할 수 있어야 한다.
        - 과거의 기준:  일간→ 요청 날짜 기준 하루 전날 부터/ 주간 → 요청 날짜 기준 전 주(일~토)
        - 즉 현재의 날짜는 달성률에 영향을 받는다.

### 새로운 테이블 추가

```sql
CREATE TABLE plan (
    plan_id BIGINT PRIMARY KEY,
    certificate_id BIGINT NOT NULL,
    member_id BIGINT NOT NULL,
    end_at DATE NOT NULL,
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);
```

```sql
CREATE TABLE plan_item (
    plan_item_id BIGINT PRIMARY KEY,
    plan_id BIGINT NOT NULL,
    goal_type ENUM('DAILY', 'WEEKLY') NOT NULL,
    activity_type ENUM('STUDY', 'EXAM', 'TIME', 'PROBLEM') NOT NULL, -- 예: STUDY_MODE, EXAM_MODE
    target_value INT NOT NULL, -- 목표 회수 또는 목표 시간
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
);
```

```sql
CREATE TABLE plan_progress (
    id BIGINT PRIMARY KEY AUTO_INCREMENT,
    plan_id BIGINT NOT NULL,
    progrpess_date DATE NOT NULL, -- 일간이면 특정 날짜, 주간이면 주의 마지막 날짜
    activity_type ENUM('STUDY', 'EXAM', 'TIME', 'PROBLEM') NOT NULL,
    goal_type ENUM('WEEKLY','DAILY') NOT NULL,
    completed_count INT NOT NULL, -- 완료된 회수 또는 시간
    target_value INT NOT NULL, -- 목표 회수 또는 목표 시간
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
);
```

### 주요 설계 고려 사항

- **과거 기록 보호** : 과거 기록을 보호하기 위해 학습 진행 시  plan_progress 에 데이터를 저장한다.
    - 플랜 수정 시 과거 기록에 영향을 받지 않는다.
- 달성률 계산 : 달력 UI 달성률 계산시 progress만 확인하기 때문에 확인해야 하는 레코드가 줄어들게 된다.
- 학습 아이템 수정에 따른 반영 :
    - 일간 목표 변경 : 해당 날짜의 plan_progress가 있으면 해당 기록을 수정한다.
    - 주간 목표 변경 : 해당 주간 범위의 plan_progress 테이블만 수정한다.

- 문제 제출시 plan_progress 테이블이 생성된다.
    - plan_progress의 최대 생성 값은 아무 데이터가 없는 경우 일간(4개), 주간(4개) 총 8개이다.
    - 이후 update 쿼리로 수행한다.

### 예상 시나리오

1. 학습 플랜 생성
    1. 사용자가 플랜 생성시 plan 테이블과 plan_item 테이블이 생성된다
    2. 날짜 범위에 따라 값이 있으면 plan_progress를 추가해야 한다.
2. 진행 상황 기록
    1. 사용자가 학습 제출시 learning 테이블에 데이터가 저장되고 이 데이터를 기반으로 
    plan_progress 테이블의 데이터를 저장하거나 수정한다.
3. 플랜 수정
    1. 플랜의 target_value값이 수정되는 경우
        1. 플랜이 수정되면 plan_items을 수정하되, plan_progress의 모든 데이터를 수정하는 것이 아닌 현재 날짜 범위에 맞는 plan_progress 값을 수정한다.(target_value)
    2. 플랜의 activity_value 값이 새로 추가가 되는 경우
        1. 새로운 plan_item을 추가한다.
        2. learning의 데이터를 기준으로 plan_item에 해당하는 값이 있는지 확인.
            1. 값이 있다면 plan_progress 테이블에 값을 추가한다. (기록을 보고 확인)
    3. 플랜의 activity_value 값이 삭제되는 경우
        1. plan_item과 날짜 범위의 plan_progress (일간,주간) 값을 삭제한다.
4. 달성률 조회
    1. plan_progress를 조회한다.
    2. 단 plan_item_history 테이블로 목표 값에 대한 더미를 만들어 주기 위해 plan_item_history 값을 읽어와 일간/ 주간 별로 값을 셋팅하고 plan_progress 값을 기반으로 달성 값을 저장한다.

### 예상되는 문제점

- plan_progress의 값을 생성 / 수정의 여부에 따라 생성하고 삭제해야 한다.
- 현재 진행중인 plan_progress가 있는지 판단을 해야 한다.
    - 일간 플랜 : 오늘 날짜
    - 주간 플랜 : 지난주 일요일 ~ 이번주 토요일

**Comment by morenow98** [comment]
plan_progress 테이블이 AUTO_INCREMENT 라고 되어 있는데, 코드에는 Tsid가 적용되어 있어요~!!
